Using attributes on fields and properties to exclude from comparison.

aksar9211 opened the pull request:
Valerii Tereshchenko, I am using object comparer in my project. I used to list out the ignore member before comparison, but In several cases, there are nested parent-child models that go into the loop recursively and result in StackOverflowException.
Also instead of excluding the member at the comparer site, It is more useful for us to specify it at the property and field. So, I am proposing the above solution for this issue. Please consider adding it to the live project.

Usage:

```
[Comparison(ComparisonStatus.ExcludeFromComparison)]
public ObservableCollection <MyClass> myObj{get; set;}

[Comparison(ComparisonStatus.ExcludeFromComparison)]
public MyClass parent{get; set;}
```

Commit:
Using attributes on fields and properties to exclude from comparison.
Valerii Tereshchenko, I am using object comparer in my project. I used list out the ignore member before comparison, but In several cases there are nested parent-child models that goes into the loop recursively and results in StackOverflowException.
Also instead of excluding the member at the comparer site, It is more useful for us to specify it at the property and field. So, I am proposing the above solution for this issue. Please consider to add it in the live project

ValeraT1982 commented on ObjectsComparer/ObjectsComparer/Enums/ComparisonStatus.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <summary>
+    /// Specifies that the element on which it is applied has to be excluded from comparison
+    /// </summary>
+    ExcludeFromComparison
```
What is the benefits of having this enum with 1 value?

aksar9211 replied:
Currently, I have added just one value in the enum as per my requirements, but further, we can extend enums to modify more settings like `IgnoreRecursiveComparison ` - if we want only one item to ignore for recursive comparison. So I am keeping my options open to manage comparison settings in one enum class.

ValeraT1982 commented on ObjectsComparer/ObjectsComparer/Attributes/ComparisonAttribute.cs:
```diff
@@ -0,0 +1,41 @@
+  /// Class is used to specify whether the element on which it is applied will have comparison effect
+  /// </summary>
+  [AttributeUsage(AttributeTargets.Property | AttributeTargets.Field)]
+  public sealed class ComparisonAttribute : Attribute
```
Consider renaming to something like IgnoreInComparison

aksar9211 replied:
Ok. That won't be a problem.

ValeraT1982 reviewed (commented):
Hi,

Thanks a lot fore the pull request. Can you please address my comments and add tests?

aksar9211 committed:
Updated Unit Testes

1. Updated the enum text as suggested.
2. Created unit tests.

aksar9211 reviewed (commented):
I have added the tests as suggested. Please review again.

ValeraT1982 commented on ObjectsComparer/ObjectsComparer/Enums/ComparisonStatus.cs:
```diff
@@ -12,8 +12,8 @@ namespace ObjectsComparer.Enums
   public enum ComparisonStatus
```
I was thinking about removing this enum completely. It simplifies things a lot.

aksar9211 replied:
Ok. I have removed it.

aksar9211 committed:
Improvement in Implementation

As per the suggestions, I have removed the comparison status enum
and renamed the ComparisonAttribute to IgnoreInComparisonAttribute.
I have also updated the unit testes.

aksar9211 commented:
As per your suggestions, I have removed the ComparisonStatus enum class and renamed the ComparisonAttribute to IgnoreInComparisonAttribute. 
I have also updated the unit testes. 
Please review again.

ValeraT1982 reviewed (commented):
Did you run all unit tests with this change?

aksar9211 commented:
Yes. I did. All the tests passed except one:
File: ComparerGenericEnumerableTests.cs  Line no.:550
Assert.AreEqual(DifferenceTypes.MissedElementInFirstObject, differences.First().DifferenceType);

the difference type is: DifferenceTypes.NumberOfElementsMismatch

This test is failing in the original project as well.

ValeraT1982 commented:
Thanks for your contribution

aksar9211 commented:
Thank you for merging the code.
How to get the nuget package for the same?

ValeraT1982 commented:
Currently the only option is to download the source code. I'm going to release next version within couple of month.

swinslade commented:
Hi, I downloaded and built the source back in June 2022 just so I could use the [IgnoreInComparison] attribute (which is working perfectly in my project), is there any chance of a new release (nuget package) for these changes in the near future? 

Thanks for the great work on this project!

ValeraT1982 commented:
Still working on it. Sorry. A lot of work.
